tools/astarte_export/lib/fetchdata/fetchdata.ex

```diff
@@ -0,0 +1,386 @@
+
+    last_credentials_request_ip =
+      device_data[:last_credentials_request_ip]
+      |> :inet_parse.ntoa()
+      |> to_string()
+
+    total_received_msgs =
+      device_data[:total_received_msgs]
+      |> to_string()
+
+    total_received_bytes =
+      device_data[:total_received_bytes]
+      |> to_string()
+
+    last_connection =
+      device_data[:last_connection]
+      |> DateTime.from_unix!(:millisecond)
+      |> DateTime.to_iso8601()
+
+    last_disconnection =
+      device_data[:last_disconnection]
+      |> DateTime.from_unix!(:millisecond)
+      |> DateTime.to_iso8601()
+
+    last_seen_ip =
+      device_data[:last_seen_ip]
+      |> :inet_parse.ntoa()
+      |> to_string()
+
+    interface_details = gen_interface_details(conn, realm, device_data)
+
+    {:ok,
+     %DeviceData{
+       device_id: device_id,
+       revision: revision,
+       pending_empty_cache: pending_empty_cache,
+       secret_bcrypt_hash: secret_bcrypt_hash,
+       first_registration: first_registration,
+       inhibit_request: inhibit_request,
+       cert_aki: cert_aki,
+       cert_serial: cert_serial,
+       first_credentials_request: first_credentials_request,
+       last_credentials_request_ip: last_credentials_request_ip,
+       total_received_msgs: total_received_msgs,
+       total_received_bytes: total_received_bytes,
+       last_connection: last_connection,
+       last_disconnection: last_disconnection,
+       last_seen_ip: last_seen_ip,
+       interfaces: interface_details
+     }}
+  end
+
+  defp gen_interface_details(conn, realm, device_data) do
+    device_id = device_data[:device_id]
+    introspection = device_data[:introspection]
+    introspection_minor = device_data[:interospection_minor]
+
+    Enum.reduce(introspection, [], fn {interface_name, major_version}, acc ->
+      {:ok, interface_description} =
+        Queries.fetch_interface_descriptor(conn, realm, interface_name, major_version)
+
+      minor_version = interface_description.minor_version |> to_string()
+      major_version1 = major_version |> Kernel.to_string()
+      interface_id = interface_description.interface_id
+      aggregation = interface_description.aggregation
+      storage = interface_description.storage
+      interface_type = interface_description.type
+      {:ok, mappings} = Queries.fetch_interface_mappings(conn, realm, interface_id)
+
+      mapped_data_fields =
+        case interface_type do
+          :datastream ->
+            case aggregation do
+              :object ->
+                fetch_object_datastreams(conn, realm, mappings, device_id, storage)
+
+              :individual ->
+                fetch_individual_datastreams(conn, realm, mappings, device_id, interface_id)
+            end
+
+          :properties ->
+            fetch_individual_properties(conn, realm, mappings, device_id, interface_id)
+        end
+
+      [
+        %{
+          interface_name: interface_name,
+          major_version: major_version1,
```

Comment:
You are assigning major_version1 several lines here above, but you just use it here.
This makes the code hard to understand.
Also why you just don't call `to_string(major_version)` here, or even better just before writing the xml file?